display_list/geometry/dl_region.cc

```diff
@@ -8,31 +8,527 @@
 
 namespace flutter {
 
-DlRegion::DlRegion(std::vector<SkIRect>&& rects) {
-  // If SpanLines can not be memmoved `addRect` would be signifantly slower
-  // due to cost of inserting and removing elements from the `lines_` vector.
-  static_assert(std::is_trivially_constructible<SpanLine>::value,
-                "SpanLine must be trivially constructible.");
-  addRects(std::move(rects));
+DlRegion::SpanBuffer::SpanBuffer(DlRegion::SpanBuffer&& m)
+    : capacity_(m.capacity_), size_(m.size_), spans_(m.spans_) {
+  m.size_ = 0;
+  m.capacity_ = 0;
+  m.spans_ = nullptr;
+};
 
-  for (auto& spanvec : spanvec_pool_) {
-    delete spanvec;
+DlRegion::SpanBuffer::~SpanBuffer() {
+  free(spans_);
+}
+
+void DlRegion::SpanBuffer::reserve(size_t capacity) {
+  if (capacity_ < capacity) {
+    spans_ = static_cast<Span*>(std::realloc(spans_, capacity * sizeof(Span)));
+    capacity_ = capacity;
+  }
+}
+
+DlRegion::SpanChunkHandle DlRegion::SpanBuffer::storeChunk(const Span* begin,
+                                                           const Span* end) {
+  size_t chunk_size = end - begin;
+  size_t min_capacity = size_ + chunk_size + 1;
+  if (capacity_ < min_capacity) {
+    size_t new_capacity = std::max(min_capacity, capacity_ * 2);
+    new_capacity = std::max(new_capacity, size_t(512));
+    reserve(new_capacity);
+  }
+  SpanChunkHandle res = size_;
+  size_ += chunk_size + 1;
+  spans_[res].left = chunk_size;
```

Comment:
Just a suggestion - add a setChunkSize method next to getChunkSize so that it is easy to verify that the same technique is used both for setting and getting?

Reply:
(I guess GetChunkSize is only for external uses?)

Reply:
Hopefully they are inlined anyway. It will be interesting if you see any performance impact, but the code would be cleaner.

I struggle with calling `left()` vs `fLeft` on SkRects. It feels quicker and simpler to just access the field, but the method feels more "API-ish compliant-ish" and compilers are very capable with inlining things.

Reply:
I couldn't see any measurable difference, I'd assume it's inlined (I'm not running the benchmarks with LTO, but few times that I tried it didn't make much difference for neither SkRegion nor DlRegion).